### PR TITLE
Enable console virtual terminal sequences processing (ANSI/VT100 colors)

### PR DESCRIPTION
**Pull request by albertony**
On Windows this makes sure console mode `ENABLE_VIRTUAL_TERMINAL_PROCESSING` is enabled on the rclone process, by using `GetConsoleMode`/`SetConsoleMode` functions from Windows API. This mode is default in many cases, e.g. when using the Windows Terminal application, but in other cases it relies on a registry setting:
```
[HKEY_CURRENT_USER\Console]
"VirtualTerminalLevel"=dword:00000001
```

(I'm still confused as to exactly how this works; when it is default enabled and not, as can be seen from the discussions in the linked PR and forum thread, but seems certain that it is not always enabled by default and therefore it is better to set it explicitly on the process).

Since rclone version 1.61.0 the tree command uses ANSI color sequences in output by default, and this lead to issues in some terminals that were not handling these (see links below).

Background:
- https://learn.microsoft.com/windows/console/console-virtual-terminal-sequences
- https://ss64.com/nt/syntax-ansi.html

#### Was the change discussed in an issue or in the forum before?

#6668
#6591
https://forum.rclone.org/t/error-with-build-v1-61-1-tree-command-panic-runtime-error-invalid-memory-address-or-nil-pointer-dereference/35922

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [ ] I have added tests for all changes in this PR if appropriate.
- [ ] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by albertony** [comment]
Not sure if this is implemented correctly with regards to the other code in the terminal package, with the TerminalColorMode variable etc, introduced in https://github.com/rclone/rclone/pull/6591. Is that related at all, or is it only for the stats/progress/config doing more direct terminal manipulations? I didn't see the terminal.Start function being called on a simple tree command.

**Comment by ncw** [comment]
> Not sure if this is implemented correctly with regards to the other code in the terminal package, with the TerminalColorMode config variable etc, introduced in #6591. I'm confused; is that related at all, or is it only for the stats/progress/config doing more direct terminal manipulations?

Hmm, not sure.

If we want to use colours on windows then we need to be calling this. Is that right?

> I didn't see the terminal.Start function being called on a simple tree command.

`Start` is called automatically if `terminal.Write()` is used.

> And when running rclone tree and piping to file `> tree.txt` in command prompt, the resulting text file does include the ansi escape codes, wasn't the intention to prevent this?

I think the problem is that tree is not using `terminal.Out` but using `os.Stdout`

I suspect if we called `terminal.Start` and used `terminal.Out` in place of `os.Stdout` in `cmd/tree/tree.go` then it would work better and maybe this patch wouldn't be needed.

Do you fancy giving it a go?

**Comment by albertony** [comment]
Hm..

> I suspect if we called `terminal.Start` and used `terminal.Out` in place of `os.Stdout` in `cmd/tree/tree.go` then it would work better

I guess that is the correct fix for this case:

> when running rclone tree and piping to file > tree.txt in command prompt, the resulting text file does include the ansi escape codes, wasn't the intention to prevent this?

I can take a look at that.

Regardless, when running tree in terminal, without piping to file, this would not make a difference for users seeing the "strange" output `←[1;34m` today, would it? I think setting the console mode explicitly, like this PR does, is something we would want to do to avoid this. So we need both fixes, for the two cases.

> If we want to use colours on windows then we need to be calling this. Is that right?

Well... To be *sure* it works we need this. In many cases it is not needed because it is already default, as controlled with the `VirtualTerminalLevel` registry value. I originally thought it was always the default in recently new versions of Windows, but it seems not. This part confuses me, but haven't had the time to dig deeper (busy times...). When I googled this before, it seemed that explicitly setting console mode was not entirely uncommon for console-oriented applications, even in recently updated ones, some of them written in go.

**Comment by ncw** [comment]
Ok!

Do you want to merge this as-is or put some more stuff on top?

**Comment by albertony** [comment]
I think I've understood the terminal package now, and added a commit which makes the tree command utilize its functionality to properly handle coloring.

TLDR;

Rclone's lib/terminal heavily relies on third party github.com/mattn/go-colorable. When colors are enabled in the terminal library, on Windows, go-colorable uses the Windows Console API. It first checks if the console mode ENABLE_VIRTUAL_TERMINAL_PROCESSING is set, and if it is then this means the console handles VT100 escape sequences natively, and go-colorable will not do anything except just relay text written to it into os.Stdout. But if the console does not have the ENABLE_VIRTUAL_TERMINAL_PROCESSING mode, then go-colorable handles the writing of text directly to the console using Console API functionality (GetConsoleScreenBufferInfo, SetConsoleTextAttribute, etc). It decodes any escape sequences found in the text, and explicitly writes the text to console instead of sending it to stdout, and are then able to set colors and other things using the API.

When colors are disabled in the terminal library, or when os.Stdout is not associated with a terminal (e.g. redirected to file), then go-colorable will strip off any escape sequences in the text, before sending the remaining text directly to the os.Stdout.

This means: Using lib/terminal for output, either through terminal.Write, which automatically calls terminal.Start(), or explicitely calling terminal.Start() and then sending text to terminal.Out intead of os.StdOut, then you should always get colored output to your Windows console - regardless of the console mode setting!

To enable native VT100 escape handling in the Windows console, the go-colorable provides utility function EnableColorsStdout which uses the Console API to set console mode ENABLE_VIRTUAL_TERMINAL_PROCESSING. Further use of lib/terminal and go-colorable will then detect this and simply send text unchanged with escape codes to os.Stdout, without adding anything extra, but writing through lib/terminal still provides the feature that colors can easily be disabled with an option that makes it strip off escape codes automatically.

Rclone now has global option `--color`, which can be set to AUTO, NEVER or ALWAYS, and this configures how lib/terminal, and then go-colorable, should treat color escape sequences as described above.

The tree command in rclone, using the third party github.com/a8m/tree, has its own option which controls if it should write color escape sequences in the first place. When using the lib/terminal this can always be enabled, since we can best let it (and go-colorable) decide what to do with the escape sequences.

**Comment by albertony** [comment]
Now my original commit, with `EnableColorsStdout` is not really needed - as you @ncw predicted:

> I suspect if we called terminal.Start and used terminal.Out in place of os.Stdout in cmd/tree/tree.go then it would work better and maybe this patch wouldn't be needed.

Still, I think perhaps it would be better. To get native Windows handling of coloring, instead of relying on the go-colorable doing a lot of stuff with the Console API. Probably faster (if it matters), and less risk of bugs/inconsistencies. Any thoughts? I will think more about it myself, a couple of days, if there are any implications etc, so if we are getting close to a release and wants the other (new) commit included I can split and merge that - just let me know.

**Review by ncw — APPROVED** [review]
LGTM if you want to merge - thank you :-)

**Comment by albertony** [comment]
Great. I've polished it a bit, with comments etc, but I think I'll just merge it as it is now.

We *could* have done the call to `terminal.EnableColorsStdout()` in the `terminal.Start` instead, to only affect "terminal-aware commands" like `tree`, instead of in the common init for all commands where it is now. But I think it is probably fine to just leave it where it is, which ensures *any* escapes emitted to stdout are being processed - on windows like it would on linux. Maybe... 😆

**Comment by ncw** [comment]
Sounds perfect :-)
